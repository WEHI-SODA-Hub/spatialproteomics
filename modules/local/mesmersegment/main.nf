process MESMERSEGMENT {
    tag "$meta.id"
    label 'process_medium_memory'

    conda "${moduleDir}/environment.yml"
    container 'ghcr.io/wehi-soda-hub/mibisegmentation:latest'

    input:
    tuple val(meta),
        path(tiff),
        val(nuclear_channel),
        val(membrane_channels),
        val(combine_method),
        val(level),
        val(maxima_threshold),
        val(interior_threshold),
        val(maxima_smooth),
        val(min_nuclei_area),
        val(remove_border_cells),
        val(pixel_expansion),
        val(padding)
    val(compartment)

    output:
    tuple val(meta), path("*.tiff"), emit: segmentation_mask
    path "versions.yml"            , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def membrane_channel_args = membrane_channels.first().split(":")
        .collect { "--membrane-channel ${it}" }.join(' ')

    // TODO: These params are a mess now, consider refactoring
    def combine_method_arg = combine_method.first() != [] ? "--combine-method ${combine_method.first()}" : ''
    def level_arg = level.first() != [] ? "--segmentation-level ${level.first()}" : ''
    def maxima_threshold_arg = maxima_threshold.first() != [] ? "--maxima-threshold ${maxima_threshold.first()}" : ''
    def interior_threshold_arg = interior_threshold.first() != [] ? "--interior-threshold ${interior_threshold.first()}" : ''
    def maxima_smooth_arg = maxima_smooth.first() != [] ? "--maxima-smooth ${maxima_smooth.first()}" : ''
    def min_nuclei_area_arg = min_nuclei_area.first() != [] ? "--min-nuclei-area ${min_nuclei_area.first()}" : ''
    def remove_border_cells_arg = remove_border_cells.first() ? '--remove-cells-touching-border' : '--no-remove-cells-touching-border'
    def pixel_expansion_arg = pixel_expansion.first() != [] ? "--pixel-expansion ${pixel_expansion.first()}" : ''
    def padding_arg = padding.first() != [] ? "--padding ${padding.first()}" : ''
    """
    mesmer-segment \\
        ${tiff} \\
        --compartment ${compartment} \\
        --nuclear-channel ${nuclear_channel.first()} \\
        ${membrane_channel_args} \\
        ${combine_method_arg} \\
        ${level_arg} \\
        ${maxima_threshold_arg} \\
        ${interior_threshold_arg} \\
        ${maxima_smooth_arg} \\
        ${min_nuclei_area_arg} \\
        ${remove_border_cells_arg} \\
        ${pixel_expansion_arg} \\
        ${padding_arg} \\
        $args \\
        > "${prefix}_${compartment}.tiff"

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        mesmersegmentation: v0.1.0
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch "${prefix}.tiff"

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        mesmersegmentation: v0.1.0
    END_VERSIONS
    """
}
