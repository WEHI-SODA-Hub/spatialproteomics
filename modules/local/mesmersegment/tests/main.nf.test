nextflow_process {

    name "Test Process MESMERSEGMENT"
    script "../main.nf"
    process "MESMERSEGMENT"

    tag "modules"
    tag "modules_nfcore"
    tag "mesmersegment"

    test("mesmersegment - nuclear segmentation") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test_nuclear' ], // meta map
                    file('${projectDir}/tests/data/mesmer/test_data.tiff', checkIfExists: true), // tiff
                    "nuclear", // nuclear_channel
                    "membrane", // membrane_channels
                    "prod", // combine_method
                    5, // level
                    0.2, // maxima_threshold
                    0.2, // interior_threshold
                    1.0, // maxima_smooth
                    100, // min_nuclei_area
                    true, // remove_border_cells
                    0, // pixel_expansion
                    64 // padding
                ]
                input[1] = "nuclear" // compartment
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("mesmersegment - cell segmentation") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test_whole_cell' ], // meta map
                    file('${projectDir}/tests/data/mesmer/test_data.tiff', checkIfExists: true), // tiff
                    "nuclear", // nuclear_channel
                    "membrane", // membrane_channels
                    "max", // combine_method
                    5, // level
                    0.05, // maxima_threshold
                    0.15, // interior_threshold
                    2.0, // maxima_smooth
                    50, // min_nuclei_area
                    false, // remove_border_cells
                    5, // pixel_expansion
                    32 // padding
                ]
                input[1] = "whole-cell" // compartment
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("mesmersegment - minimal parameters") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test_minimal' ], // meta map
                    file('${projectDir}/tests/data/mesmer/test_data.tiff', checkIfExists: true), // tiff
                    "nuclear", // nuclear_channel
                    "membrane", // membrane_channels
                    null, // combine_method
                    null, // level
                    null, // maxima_threshold
                    null, // interior_threshold
                    null, // maxima_smooth
                    null, // min_nuclei_area
                    false, // remove_border_cells
                    null, // pixel_expansion
                    null // padding
                ]
                input[1] = "nuclear" // compartment
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("mesmersegment - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [ id:'test_stub' ], // meta map
                    file('${projectDir}/tests/data/mesmer/test_data.tiff', checkIfExists: true), // tiff
                    "nuclear", // nuclear_channel
                    "membrane", // membrane_channels
                    "prod", // combine_method
                    5, // level
                    0.1, // maxima_threshold
                    0.2, // interior_threshold
                    1.0, // maxima_smooth
                    100, // min_nuclei_area
                    true, // remove_border_cells
                    0, // pixel_expansion
                    64 // padding
                ]
                input[1] = "nuclear" // compartment
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
