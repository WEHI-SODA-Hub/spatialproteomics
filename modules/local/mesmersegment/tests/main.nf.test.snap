{
    "mesmersegment - nuclear segmentation": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": [
                                "analyse_samplesheet.csv:md5,bf995bff1feac8b4fc2b4f46e7ae5b08"
                            ]
                        },
                        "test.tiff:md5,61e1948c868ea978e60b1ed77a65fe7f"
                    ]
                ],
                "1": [
                    "versions.yml:md5,295760d34b0dcd2c396b9fa080140d7e"
                ],
                "segmentation_mask": [
                    [
                        {
                            "id": [
                                "analyse_samplesheet.csv:md5,bf995bff1feac8b4fc2b4f46e7ae5b08"
                            ]
                        },
                        "test.tiff:md5,61e1948c868ea978e60b1ed77a65fe7f"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,295760d34b0dcd2c396b9fa080140d7e"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-26T14:16:33.718304447"
    },
    "mesmersegment - cell segmentation": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_cell"
                        },
                        "test_cell.tiff:md5,a6a68d467c2156e78ee578bf4991ee84"
                    ]
                ],
                "1": [
                    "versions.yml:md5,295760d34b0dcd2c396b9fa080140d7e"
                ],
                "segmentation_mask": [
                    [
                        {
                            "id": "test_cell"
                        },
                        "test_cell.tiff:md5,a6a68d467c2156e78ee578bf4991ee84"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,295760d34b0dcd2c396b9fa080140d7e"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-26T14:17:12.121046827"
    },
    "mesmersegment - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_stub"
                        },
                        "test_stub.tiff:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,295760d34b0dcd2c396b9fa080140d7e"
                ],
                "segmentation_mask": [
                    [
                        {
                            "id": "test_stub"
                        },
                        "test_stub.tiff:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,295760d34b0dcd2c396b9fa080140d7e"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-26T14:18:29.734892681"
    },
    "mesmersegment - minimal parameters": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_minimal"
                        },
                        "test_minimal.tiff:md5,7f9de8e817dff7ca8b7544a3a5c37426"
                    ]
                ],
                "1": [
                    "versions.yml:md5,295760d34b0dcd2c396b9fa080140d7e"
                ],
                "segmentation_mask": [
                    [
                        {
                            "id": "test_minimal"
                        },
                        "test_minimal.tiff:md5,7f9de8e817dff7ca8b7544a3a5c37426"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,295760d34b0dcd2c396b9fa080140d7e"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-06-26T14:18:00.860492248"
    }
}